File: Round 2/bilingual.py
# Copyright (c) 2015 kamyu. All rights reserved.
#
# Google Code Jam 2015 Round 2 - Problem C. Bilingual
# https://code.google.com/codejam/contest/8234486/dashboard#s=p2
#
# Time  : O((N * L)^2)
# Space : O(N * L)
#

def dfs(node, sink, used, E):
    if used[node]:
        return False
    used[node] = True

    for i in xrange(len(E[node])):
        if E[node][i] == sink or dfs(E[node][i], sink, used, E):
            E[E[node][i]].append(node)
            E[node][-1], E[node][i] = E[node][i], E[node][-1]
            E[node].pop()
            return True
    return False

def bilingual():
    N = input()
    
    def word_id(word_ids, word):
        if word in word_ids:
            return word_ids[word]
        word_ids[word] = len(word_ids)
        return word_ids[word]

    # Parse lines.
    word_ids = {}
    lines = [list(set([word_id(word_ids, word) \
             for word in raw_input().strip().split()])) \
             for _ in xrange(N)]

    # Init edges.
    # i (0 ~ N) node represents the ith line.
    # 2 * i + N node represents the word i is in English.
    # 2 * i + N + 1 node represents the word i is in French.
    source, sink = 0, 1
    E = [[] for _ in xrange(2 * len(word_ids) + N)]
    for i in xrange(len(word_ids)):
        E[2 * i + N].append(2 * i + N + 1)
    for x in lines[0]:
        E[source].append(2 * x + N)
    for y in lines[1]:
        E[2 * y + N + 1].append(sink)
    for i in xrange(2, N):
        for x in lines[i]:
            E[2 * x + N + 1].append(i)
            E[i].append(2 * x + N)
    # Run max flow.
    flow = 0
    used = [False for _ in xrange(len(E))]
    while dfs(source, sink, used, E):
        flow += 1
        used = [False for _ in xrange(len(E))]
    return flow


for case in xrange(input()):
    print "Case #%d: %d" % (case + 1, bilingual())

File: Round 2/bilingual.cpp
// Copyright (c) 2015 kamyu. All rights reserved.

/*
 * Google Code Jam 2015 Round 2 - Problem C. Bilingual
 * https://code.google.com/codejam/contest/8234486/dashboard#s=p2
 *
 * Time  : O(N * W)
 * Space : O(N + W)
 *
 */

#include <iostream>
#include <string>
#include <sstream>
#include <vector>
#include <unordered_map>
#include <unordered_set>
#include <algorithm>

using std::cin;
using std::cout;
using std::endl;
using std::istringstream;
using std::string;
using std::vector;
using std::unordered_map;
using std::unordered_set;
using std::swap;

bool dfs(int node, int sink,
        vector<bool> *used_ptr,
         vector<vector<int>> *E_ptr) {
    vector<bool>& used = *used_ptr;
    vector<vector<int>>& E = *E_ptr;

    if (used[node]) {
        return false;
    }
    used[node] = true;
    for (auto it = E[node].begin(); it != E[node].end(); ++it) {
        auto x = *it;
        if (x == sink || dfs(x, sink, &used, &E)) {
            E[x].emplace_back(node);
            swap(E[node].back(), *it);
            E[node].pop_back();
            return true;
        }
    }
    return false;
}

int bilingual() {
    int N;
    cin >> N;

    string ss; getline(cin, ss);
    unordered_map<string, int> values;
    auto get = [&values](string s) {
        if (values.count(s)) {
            return values[s];
        }
        return (values[s] = values.size());
    };

    // Parse lines.
    vector<unordered_set<int>> lines(N);
    for (int i = 0; i < N; ++i) {
        string line; getline(cin, line);
        istringstream in(line);
        string word;
        while (in >> word) {
            lines[i].emplace(get(word));
        }
    }

    // Init edges.
    int source = 0, sink = 1;
    vector<vector<int>> E(2 * values.size() + N);
    for (int i = 0; i < values.size(); ++i) {
        E[2 * i + N].emplace_back(2 * i + N + 1);
    }
    for (const auto &x : lines[0]) {
        E[source].emplace_back(2 * x + N);
    }
    for (const auto &y : lines[1]) {
        E[2 * y + N + 1].emplace_back(sink);
    }
    for (int i = 2; i < N; ++i) {
        for (const auto &x : lines[i]) {
            E[2 * x + N + 1].emplace_back(i);
            E[i].emplace_back(2 * x + N);
        }
    }

    // Run max flow.
    int flow = 0;
    vector<bool> used(E.size(), false);
    while (dfs(source, sink, &used, &E)) {
        ++flow;
        used = move(vector<bool>(E.size(), false));
    }
    return flow;
}

int main() {
    int T;
    cin >> T;

    for (int test = 1; test <= T; ++test) {
        cout << "Case #" << test << ": " << bilingual() << endl;
    }
}
